src/main/java/com/vmware/vipclient/i18n/messages/service/StringService.java

```diff
@@ -16,37 +16,64 @@
 import com.vmware.vipclient.i18n.messages.api.opt.server.StringBasedOpt;
 import com.vmware.vipclient.i18n.messages.dto.MessagesDTO;
 import com.vmware.vipclient.i18n.util.ConstantsKeys;
 import com.vmware.vipclient.i18n.util.JSONUtils;
 
 public class StringService {
-    private MessagesDTO dto    = null;
-
     Logger              logger = LoggerFactory.getLogger(StringService.class);
-
-    public StringService(MessagesDTO dto) {
-        this.dto = dto;
-    }
-
+    
     @SuppressWarnings("unchecked")
-    public String getString() {
-        String key = dto.getKey();
-        CacheService cacheservice = new CacheService(dto);
-        Map<String, String> map = cacheservice.getCacheOfComponent();
-        if (map == null) {
-            if (!cacheservice.isContainComponent()) {
-                Object o = new ComponentService(dto).getMessages();
-                map = (Map<String, String>) o;
-                cacheservice.addCacheOfComponent(map);
-            }
-        }
-        return (map == null || map.get(key) == null ? "" : map.get(key));
+    public String getString(MessagesDTO dto) {
+    	String key = dto.getKey();
+    	CacheService cacheService = new CacheService(dto);
+    	CacheItem cacheItem = cacheService.getCacheOfComponent();
+    	Map<String, String> cacheOfComponent = null;
+    	if (cacheItem != null) { // Item is in cache
+    		cacheOfComponent = cacheItem.getCachedData();
+    		if (cacheService.isExpired()) { // cacheItem has expired
+    			// Update the cache in a separate thread
+    			populateCacheTask(cacheItem.getCacheProperties(), cacheService, dto); 		
+    		}
+    	} else if ((cacheItem == null && !cacheService.isContainComponent())) { // Item is not in cache
+    		// Create a new HashMap to store cache properties.
+    		cacheOfComponent = populateCache(new HashMap<String, Object>(), cacheService, dto);
+       } 
+       return (cacheOfComponent == null || cacheOfComponent.get(key) == null ? "" : cacheOfComponent.get(key));
+    }
+    
+    private volatile boolean running = true;
+	private void populateCacheTask(Map<String, Object> cacheProps, final CacheService cacheService, MessagesDTO dto) {
+		Runnable task = () -> {
+	    	while (running) {
+	    		try {
+			    	// Use the cacheProps that is already in the cache.
+			    	populateCache(cacheProps, cacheService, dto);
+	    		} finally {
+			    	running = false;
+			    }
+	    	}
+		    
+		};
+		new Thread(task).start();
+	}
+	
+	private Map<String, String> populateCache(Map<String, Object> cacheProps, 
+			CacheService cacheService, MessagesDTO dto) {
+    	// Pass cacheProps to getMessages so that:
+		// 1. A previously stored ETag, if any, can be used for the next HTTP request.
+		// 2. Cached properties can be refreshed with new properties from the next HTTP response.
+		Object o = new ComponentService(dto).getMessages(cacheProps);
+		
+		Map<String, String>  cacheOfComponent = (Map<String, String>) o;
```

Comment:
Suggest changing to Map<String, String>  cacheOfComponent = new ComponentService(dto).getMessages(cacheProps);